File: exercises/floor03/project_loot.py
"""
FLOOR 3 · PROJECT — THE LOOT TABLE
Third organ of the roguelike: loot and a stacking inventory. When the
procgen floor generator arrives on Floor 7, it will pull from this table.

Run:  python3 dojo.py check 3-P
Save: python3 dojo.py done 3-P
"""

# Roll a d100. You get the FIRST entry whose max_roll is >= your roll.
LOOT_TABLE = [
    {"max_roll": 50, "item": "gold coin", "rarity": "common"},
    {"max_roll": 75, "item": "torch", "rarity": "common"},
    {"max_roll": 90, "item": "protein bar", "rarity": "uncommon"},
    {"max_roll": 99, "item": "steel sword", "rarity": "rare"},
    {"max_roll": 100, "item": "glowstick of the ancients", "rarity": "legendary"},
]


def roll_loot(roll):
    """Given a roll from 1-100, return the ITEM NAME won.
    Loop through LOOT_TABLE in order; first entry with max_roll >= roll wins.
        roll_loot(3)   -> "gold coin"
        roll_loot(76)  -> "protein bar"
        roll_loot(100) -> "glowstick of the ancients"
    """
    # TODO
    ...


def add_to_pack(pack, item):
    """The pack is a STACKING inventory — a dict of item -> count.
    Add one of `item` and return the pack.
        add_to_pack({}, "torch")            -> {"torch": 1}
        add_to_pack({"torch": 2}, "torch")  -> {"torch": 3}
    (That .get(key, 0) pattern from exercise 3-3 pays off immediately.)
    """
    # TODO
    ...


def open_chest(pack, rolls):
    """A chest gives several rolls at once. For each roll in `rolls`,
    roll the loot and add it to the pack. Return the pack.
        open_chest({}, [3, 76, 3]) -> {"gold coin": 2, "protein bar": 1}
    Compose your two functions — write no new logic.
    """
    # TODO
    ...


def pack_report(pack):
    """Return the pack as display lines, ALPHABETICAL by item, one string:
        pack_report({"torch": 2, "gold coin": 5})
        -> "gold coin x5\ntorch x2"
    sorted(pack) gives you the keys in order. Join lines with "\n".
    Empty pack -> "(empty)"
    """
    # TODO
    ...

File: exercises/cpp/floor03/project_loot.cpp
// FLOOR C3 · PROJECT — THE LOOT TABLE (C++ edition)
// Same d100 table, same stacking pack — with a gift at the end: std::map
// keeps its keys sorted automatically, so the alphabetized report that took
// Python a sorted() call falls out for free.
//
// Run:  python3 dojo.py check c3-P
// Save: python3 dojo.py done c3-P

#include <string>
#include <vector>
#include <map>

struct LootEntry {
    int max_roll;
    std::string item;
};

std::vector<LootEntry> LOOT_TABLE = {
    {50,  "gold coin"},
    {75,  "torch"},
    {90,  "protein bar"},
    {99,  "steel sword"},
    {100, "glowstick of the ancients"},
};

// d100 roll: first entry whose max_roll >= roll wins.
std::string roll_loot(int roll) {
    // TODO
    return "";
}

// Stacking pack: item -> count. pack[item]++ does everything.
void add_to_pack(std::map<std::string, int>& pack, std::string item) {
    // TODO
}

// For each roll, roll_loot then add_to_pack. Compose — no new logic.
void open_chest(std::map<std::string, int>& pack, const std::vector<int>& rolls) {
    // TODO
}

// "gold coin x5\ntorch x2" — alphabetical. std::map iterates in key order
// already; just build the lines. Empty pack -> "(empty)"
std::string pack_report(const std::map<std::string, int>& pack) {
    // TODO
    return "";
}
